Firmware/RTK_Surveyor/Tasks.ino

```diff
@@ -50,191 +52,222 @@ void F9PSerialReadTask(void *e)
-        //We'll wrap next loop
-        if ((btTail + btBytesToSend) >= sizeof(rBuffer))
-          btBytesToSend = sizeof(rBuffer) - btTail;
-
-        //Push new data to BT SPP if not congested or not throttling
-        btBytesToSend = bluetoothWriteBytes(&rBuffer[btTail], btBytesToSend);
-        if (btBytesToSend > 0)
-        {
-          //If we are in base mode, assume part of the outgoing data is RTCM
-          if (systemState >= STATE_BASE_NOT_STARTED && systemState <= STATE_BASE_FIXED_TRANSMITTING)
-            bluetoothOutgoingRTCM = true;
-        }
-        else
-          log_w("BT failed to send");
+      //Account for the new data
+      if (btConnected)
+        btBytesToSend += newBytesToRecord;
+      if (online.logging)
+        sdBytesToRecord += newBytesToRecord;
+      if (online.nmeaServer && wifiNmeaConnected)
+        nmeaBytesToSend += newBytesToRecord;
+    } //End Serial.available()
+
+    //----------------------------------------------------------------------
+    //Send data over Bluetooth
+    //----------------------------------------------------------------------
 
+    //If we are actively survey-in then do not pass NMEA data from ZED to phone
+    if (!btConnected)
+      //Discard the data
+      btTail = dataHead;
+    else
+    {
+      //Reduce bytes to send if we have more to send then the end of the buffer
+      //We'll wrap next loop
+      if ((btTail + btBytesToSend) >= sizeof(rBuffer))
+        btBytesToSend = sizeof(rBuffer) - btTail;
+
+      //Push new data to BT SPP if not congested or not throttling
+      btBytesToSend = bluetoothWriteBytes(&rBuffer[btTail], btBytesToSend);
+      if (btBytesToSend > 0)
+      {
+        //If we are in base mode, assume part of the outgoing data is RTCM
+        if (systemState >= STATE_BASE_NOT_STARTED && systemState <= STATE_BASE_FIXED_TRANSMITTING)
+          bluetoothOutgoingRTCM = true;
 
         //Account for the sent or dropped data
         btTail += btBytesToSend;
         if (btTail >= sizeof(rBuffer))
           btTail -= sizeof(rBuffer);
       }
+      else
+        log_w("BT failed to send");
+    }
 
-      //----------------------------------------------------------------------
-      //Log data to the SD card
-      //----------------------------------------------------------------------
+    //----------------------------------------------------------------------
+    //Send data to the NMEA clients
+    //----------------------------------------------------------------------
 
-      //If user wants to log, record to SD
-      if (!online.logging)
-        //Discard the data
-        sdTail = dataHead;
-      else
+    if (!settings.enableNmeaServer && (!wifiNmeaConnected))
+      nmeaTail = dataHead;
+    else
+    {
+      //Reduce bytes to send if we have more to send then the end of the buffer
+      //We'll wrap next loop
+      if ((nmeaTail + nmeaBytesToSend) > sizeof(rBuffer))
```

Comment:
I think this > sizeof() is correct (as opposed to >= sizeof()) but we should correct it in the other two spots (SD and BT)